src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/yflows/YFlowRerouteSpec.groovy

```diff
@@ -141,4 +150,197 @@ class YFlowRerouteSpec extends HealthCheckSpecification {
+        def yFlowRequest = yFlowHelper.randomYFlow(swT, false)
+        def yFlow = yFlowHelper.addYFlow(yFlowRequest)
+        def yFlowPathBeforeReroute = northboundV2.getYFlowPaths(yFlow.YFlowId)
+
+        when: "Y-Flow reroute has been called"
+        northboundV2.rerouteYFlow(yFlow.YFlowId)
+
+        then: "The appropriate error has been returned"
+        def actualException = thrown(HttpClientErrorException)
+        new YFlowRerouteExpectedError(~/Reroute is unsuccessful. Couldn't find new path\(s\)/).matches(actualException)
+        actualException.message.contains("The flow already on the best path")
+
+        and: "Y-Flow path has not been changed"
+        def yFlowPathAfterReroute = northboundV2.getYFlowPaths(yFlow.YFlowId)
+        verifyAll {
+            northboundV2.getYFlow(yFlow.YFlowId).status == FlowState.UP.toString()
+            yFlowPathAfterReroute == yFlowPathBeforeReroute
+        }
+
+        cleanup:
+        yFlow && yFlowHelper.deleteYFlow(yFlow.YFlowId)
+    }
+
+    @Tidy
+    @Tags([LOW_PRIORITY])
+    def "Y-Flow reroute has been executed when more preferable path is available for both sub-flows (shared path cost was changed)" () {
+        given: "The appropriate switches have been collected"
+        //y-flow with shared path is created when shared_ep+ep1->neighbour && ep1+ep2->neighbour && shared_ep+ep2->not neighbour
+        def pairSharedEpAndEp1 = topologyHelper.getAllSwitchPairs().neighbouring().first()
+        def sharedEpNeighbouringSwitches = topologyHelper.getAllSwitchPairs().neighbouring().includeSwitch(pairSharedEpAndEp1.src).collectSwitches()
+        def pairEp1AndEp2 = topologyHelper.getAllSwitchPairs().neighbouring().excludePairs([pairSharedEpAndEp1])
+                .includeSwitch(pairSharedEpAndEp1.dst).excludeSwitches(sharedEpNeighbouringSwitches).first()
+        def swT = new SwitchTriplet(shared: pairSharedEpAndEp1.src, ep1: pairEp1AndEp2.src, ep2: pairEp1AndEp2.dst)
+
+        and: "The ISLs cost between switches has been changed to make preferable path"
+        List<Isl> directSwTripletIsls = (pairSharedEpAndEp1.paths + pairEp1AndEp2.paths).findAll { it.size() == 2 }
+                .collectMany { pathHelper.getInvolvedIsls(it) }
+
+        northbound.updateLinkProps(directSwTripletIsls.collectMany { isl ->
+            [islUtils.toLinkProps(isl, ["cost": (0).toString()])]
+        })
+
+        and: "Y-Flow with shared path has been created successfully"
+        def yFlow = yFlowHelper.addYFlow(yFlowHelper.randomYFlow(swT, false))
+        def yFlowPathBeforeReroute = northboundV2.getYFlowPaths(yFlow.YFlowId)
+
+        and: "Shared ISLs cost has been changed to provide on-demand Y-Flow reroute"
+        def sharedPathIslBeforeReroute = new Path(yFlowPathBeforeReroute.sharedPath.forward + yFlowPathBeforeReroute.sharedPath.reverse, topology).getInvolvedIsls()
+        northbound.updateLinkProps(sharedPathIslBeforeReroute.collectMany { isl ->
```

Comment:
Can we move this to some helper or IslExtended class?

Reply:
Done.